test/NuGet.Server.Tests/IntegrationTests.cs

```diff
@@ -139,31 +148,103 @@ public async Task PushPackageThenReadPackages()
 
                 // Act & Assert
                 // 1. Push the package.
-                using (var request = new HttpRequestMessage(HttpMethod.Put, "/nuget")
+                await tc.PushPackageAsync(apiKey, packagePath);
+
+                // 2. Get the list of packages. This should mention the package.
+                using (var request = new HttpRequestMessage(HttpMethod.Get, "/nuget/Packages()"))
+                using (var response = await tc.Client.SendAsync(request))
                 {
-                    Headers =
-                    {
-                        { "X-NUGET-APIKEY", apiKey }
-                    },
-                    Content = tc.GetFileUploadContent(packagePath)
-                })
+                    Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+                    var content = await response.Content.ReadAsStringAsync();
+
+                    Assert.Contains(TestData.PackageId, content);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Added due to https://github.com/NuGet/NuGetGallery/issues/6960. There was a concurrency issue when pushing
+        /// packages that could lead to unnecessary cache rebuilds.
+        /// </summary>
+        [Fact]
+        public async Task DoesNotRebuildTheCacheWhenPackagesArePushed()
+        {
+            // Arrange
+            using (var tc = new TestContext(_output))
+            {
+                // Essentially disable the automatic cache rebuild by setting it really far in the future.
+                const int initialCacheRebuildAfterSeconds = 60 * 60 * 24;
+                tc.Settings["initialCacheRebuildAfterSeconds"] = initialCacheRebuildAfterSeconds.ToString();
+
+                const int workerCount = 8;
+                const int totalPackages = 80;
+
+                var packagePaths = new ConcurrentBag<string>();
+                for (var i = 0; i < totalPackages; i++)
                 {
-                    using (request)
-                    using (var response = await tc.Client.SendAsync(request))
+                    var packageId = Guid.NewGuid().ToString();
+                    var packagePath = Path.Combine(tc.TemporaryDirectory, $"{packageId}.1.0.0.nupkg");
+                    using (var package = TestData.GenerateSimplePackage(packageId, SemanticVersion.Parse("1.0.0")))
+                    using (var fileStream = File.OpenWrite(packagePath))
```

Comment:
Does anything clean up these packages? If not, consider using `FileOptions.DeleteOnClose`

Reply:
The test context has a temp dir that cleans this up.